### PR TITLE
[stdlib][do not merge] Zip collection

### PR DESCRIPTION
**Pull request by airspeedswift**
Adds collection-conforming `Zip2Collection`, and conditionally conforms it to `RandomAccessCollection` and `Equatable` when both underlying collections are.

Needs evolution proposal.

### CODE REVIEW

**Comment by airspeedswift** [comment]
@swift-ci please smoke test compiler performance

**Comment by airspeedswift** [comment]
@swift-ci please smoke benchmark

**Comment by airspeedswift** [comment]
@swift-ci please test

File: stdlib/public/core/Zip.swift [hunk 1]
```diff
@@ -82,9 +82,11 @@ public struct Zip2Sequence<Sequence1 : Sequence, Sequence2 : Sequence> {
   /// Creates an instance that makes pairs of elements from `sequence1` and
   /// `sequence2`.
   @_inlineable // FIXME(sil-serialize-all)
-  public // @testable
-  init(_sequence1 sequence1: Sequence1, _sequence2 sequence2: Sequence2) {
-    (_sequence1, _sequence2) = (sequence1, sequence2)
+  @_versioned
+  internal // @testable
+  init(_ sequence1 : Sequence1, _ sequence2: Sequence2) {
```
Comment (xwu) [inline stdlib/public/core/Zip.swift]: Drive-by nit: extraneous space before first colon.

**Comment by airspeedswift** [comment]
@swift-ci please test

**Comment by airspeedswift** [comment]
@swift-ci please test

File: stdlib/public/core/Zip.swift [hunk 2]
```diff
@@ -153,5 +151,261 @@ extension Zip2Sequence: Sequence {
...
+
+public struct Zip2Collection<Collection1 : Collection, Collection2 : Collection> {
+  @_versioned // FIXME(sil-serialize-all)
+  internal let _collection1: Collection1
+  @_versioned // FIXME(sil-serialize-all)
+  internal let _collection2: Collection2
+  
+  /// Creates an instance that makes pairs of elements from `collection1` and
+  /// `collection2`.
+  // FIXME: this should be inlineable, but this results in a compilation error:
```
Comment (slavapestov) [inline stdlib/public/core/Zip.swift]: When a non-delegating initializer is inlinable, it exposes the stored property layout of the type. So the type has to be @_fixed_layout, which this one is not.

Comment (airspeedswift) [reply stdlib/public/core/Zip.swift]: Ahhhh thanks that's what I was missing.

I raised [SR-6759](https://bugs.swift.org/browse/SR-6759) for a clearer diagnostic.

**Comment by airspeedswift** [comment]
@swift-ci please test

**Comment by airspeedswift** [comment]
@swift-ci please benchmark

**Comment by airspeedswift** [comment]
Disk space failure on the macOS tests.

@swift-ci please test macOS platform

File: validation-test/stdlib/Zip.swift [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+    indicesType: DefaultIndices<Collection>.self)
+
+  for test in zipTests {
+    let left = MinimalRandomAccessCollection<OpaqueValue<Int>>(
+      elements: test.sequence.map(OpaqueValue.init))
+    let right = MinimalRandomAccessCollection<OpaqueValue<Int32>>(
+      elements: test.other.map(OpaqueValue.init))
+
+    var result = zip(left, right)
+    expectType(Collection.self, &result)
```
Comment (moiseev) [inline validation-test/stdlib/Zip.swift]: `expectType(RandomAccessCollection.self, &result)`

File: validation-test/stdlib/Zip.swift [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+  expectTrue(zip(0..<0,0..<0).isEmpty)
+  expectTrue(zip(0..<1,0..<0).isEmpty)
+  expectTrue(zip(0..<0,0..<1).isEmpty)
+  expectFalse(zip(0..<1,0..<1).isEmpty)
+  expectFalse(zip(0..<2,0..<3).isEmpty)
+  expectFalse(zip(0..<3,0..<2).isEmpty)
+}
+
+ZipTests.test("Collection.count") {
+  expectEqual(zip(0..<0,0..<0).underestimatedCount, 0)
```
Comment (moiseev) [inline validation-test/stdlib/Zip.swift]: It turns out we never define `underestimatedCount` for `Zip2Sequence`. I think it should be `Swift.min(seq1.underestimatedCount, seq2.underestimatedCount)`.

Comment (moiseev) [reply validation-test/stdlib/Zip.swift]: Also probably worth overriding `underestimatedCount` for `Zip2Collection` as well, returning the minimum of underestimated counts of the base collections, and only calling `count` for the `RandomAccessCollection` case.

File: validation-test/stdlib/Zip.swift [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+  expectEqual(zip(0..<0,0..<0).count, 0)
+  expectEqual(zip(0..<1,0..<0).count, 0)
+  expectEqual(zip(0..<0,0..<1).count, 0)
+  expectEqual(zip(0..<2,0..<1).count, 1)
+  expectEqual(zip(0..<1,0..<2).count, 1)
+  expectEqual(zip(0..<2,0..<2).count, 2)
+}
+
+ZipTests.test("Equatable") {
+  let s =
```
Comment (moiseev) [inline validation-test/stdlib/Zip.swift]: Does his compile?!

Comment (airspeedswift) [reply validation-test/stdlib/Zip.swift]: Huh. Yeah, with a warning that lit doesn't surface. Good spot...

File: validation-test/stdlib/Zip.swift (hunk 3 above)
Comment (airspeedswift) [reply validation-test/stdlib/Zip.swift]: Don't think so. That's `Collection`, the type alias a few lines above, rather than `Collection` the protocol. Probably best to rename that alias...

File: validation-test/stdlib/Zip.swift (hunk 4 above)
Comment (airspeedswift) [reply validation-test/stdlib/Zip.swift]: Yeah good point, will add that.

File: validation-test/stdlib/Zip.swift (hunk 5 above)
Comment (moiseev) [reply validation-test/stdlib/Zip.swift]: Ah.. Right. You're producing an `s: Unit`.

**Comment by airspeedswift** [comment]
@swift-ci please test
